### PR TITLE
Support EQ and NE for array and struct

### PR DESCRIPTION
**Pull request by mx990**

### CODE REVIEW

File: src/core/datatypes/forte_array_common.h [hunk 1]
```diff
@@ -29,7 +29,24 @@
  */
 template<typename T>
 class CIEC_ARRAY_COMMON : public CIEC_ARRAY {
-public:
+    friend bool operator==(const CIEC_ARRAY_COMMON &paLeft, const CIEC_ARRAY_COMMON &paRight) {
+      if (paLeft.getLowerBound() != paRight.getLowerBound() || paLeft.getUpperBound() != paRight.getUpperBound()) {
+        return false;
+      }
+
+      for (intmax_t i = paLeft.getLowerBound(), end = paRight.getUpperBound(); i <= end; ++i) {
```
Comment (MartinMelikMerkumians) [inline src/core/datatypes/forte_array_common.h]: I know that the bounds of left and right are checked on equality before, but is there a special reason to use paLeft.getLowerBound() and then paRight.getUpperBound()?
If there is no reason, I would personally prefer to use bounds of either left or right, as people might wonder if there is a special reason

File: src/core/datatypes/forte_array.h [hunk 2]
```diff
@@ -27,7 +27,24 @@
 /** \brief A common supertype for all CIEC_ARRAY variants, providing the minimal interface an array must provide
  */
 class CIEC_ARRAY : public CIEC_ANY_DERIVED {
-public:
+    friend bool operator==(const CIEC_ARRAY &paLeft, const CIEC_ARRAY &paRight) {
+      if (paLeft.getLowerBound() != paRight.getLowerBound() || paLeft.getUpperBound() != paRight.getUpperBound()) {
+        return false;
+      }
+
+      for (intmax_t i = paLeft.getLowerBound(), end = paRight.getUpperBound(); i <= end; ++i) {
```
Comment (MartinMelikMerkumians) [inline src/core/datatypes/forte_array.h]: I know that the bounds of left and right are checked on equality before, but is there a special reason to use paLeft.getLowerBound() and then paRight.getUpperBound()?
If there is no reason, I would personally prefer to use bounds of either left or right, as people might wonder if there is a special reason

**Review by MartinMelikMerkumians — CHANGES_REQUESTED** [review]

File: src/core/datatypes/forte_array_common.h (hunk 1 above)
Comment (mx990) [reply src/core/datatypes/forte_array_common.h]: Ah, makes sense. I changed it.

File: src/core/datatypes/forte_array.h (hunk 2 above)
Comment (mx990) [reply src/core/datatypes/forte_array.h]: Ah, makes sense. I changed it.

**Review by MartinMelikMerkumians — APPROVED** [review]

**Pull request merged**